File: 350. Intersection of Two Arrays II/intersect.py
from typing import List

class Solution:
  def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
    nums1Map = {}
    res = []
    for n in nums1:
      if n not in nums1Map:
        nums1Map[n] = 0
      nums1Map[n] += 1
        
    for n in nums2:
      if n in nums1Map and nums1Map[n] > 0:
        res.append(n)
        nums1Map[n] -= 1
        
    return res


nums1 = [1, 2, 2, 1]
nums2 = [2, 2]
solution = Solution()
print('intersect: ', solution.intersect(nums1, nums2))

File: 350. Intersection of Two Arrays II/intersect.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
const intersect = (nums1, nums2) => {
  const nums1Map = {};
  const res = [];
  for (let i = 0; i < nums1.length; i++) {
    if (!nums1Map[nums1[i]]) {
      nums1Map[nums1[i]] = 0;
    }
    nums1Map[nums1[i]]++;
  }

  for (let i = 0; i < nums2.length; i++) {
    if (nums1Map[nums2[i]] && nums1Map[nums2[i]] > 0) {
      res.push(nums2[i]);
      nums1Map[nums2[i]]--;
    }
  }

  return res;
};


const nums1 = [1, 2, 2, 1];
const nums2 = [2, 2];
console.log('intersect: ', intersect(nums1, nums2));
